[Docs] Add Documents for DataElements (#1381)
open-mmlab/mmocr: docs/en/basic_concepts/structures.md

| padded_indexes | `torch.LongTensor` | If the length of indexes is less than the maximum sequence length and `pad_idx` exists, this field holds the encoded text sequence padded to the maximum sequence length of `max_seq_len`. |

## DataSample xxxDataSample

By defining a uniform data structure, we can easily encapsulate the annotation data and prediction results in a uniform way, making data transfer between different modules of the code base easier. In MMOCR, we have encapsulated three data abstractions based on the three tasks we now support and the data they need, including the text detection data abstraction [`TextDetDataSample`](mmocr.structures.TextDetDataSample), the text recognition data abstraction [`TextRecogDataSample`](mmocr.structures.TextRecogDataSample), and the key information extraction data abstraction [`KIEDataSample`](mmocr.structures.KIEDataSample). These data abstractions all inherit from {external+mmengine:doc}`MMEngine: data base class <advanced_tutorials/data_element>` `BaseDataElement`, which is used to hold all annotation and prediction information required by each task.

### Text Detection Data Abstraction TextDetDataSample

> [TextDetDataSample](mmocr.structures.TextDetDataSample) is used to encapsulate the data needed for the text detection task. It contains two main fields `gt_instances` and `pred_instances`, which are used to store the prediction results and annotation information respectively.

Reviewer: ```suggestion
[TextDetDataSample](mmocr.structures.TextDetDataSample) is used to encapsulate the data needed for the text detection task. It contains two main fields `gt_instances` and `pred_instances`, which are used to store the annotation information and prediction results respectively.
```

Reviewer: Swap the order in CN version

|                |                                 |                         |
| -------------- | ------------------------------- | ----------------------- |
| Field          | Type                            | Description             |
| gt_instances   | [`InstanceData`](#instancedata) | Annotation information. |

> | pred_instances | [`InstanceData`](#instancedata) | The predicted result.   |

Reviewer: fix links

Reviewer: ```suggestion
| gt_instances   | [`InstanceData`](#instancedata) | Annotation information. |
| pred_instances | [`InstanceData`](#instancedata) | Prediction results.   |
```

pred_instances.bboxes = torch.rand((5, 4))
pred_instances.labels = torch.zeros((5,), dtype=torch.long)
data_sample.pred_instances = pred_instances
```

[`TextDetDataSample`](mmocr.structures.TextRecogDataSample) is the basic data structure used for model inference and training in text detection tasks. The following code shows the flow of [`TextDetDataSample`](mmocr.structures.TextRecogDataSample) between the modules of the model, using DBNet as an example.

```python
# Step 1: DBHead, input the initial image img and data_samples from the data pipeline, and return the output of the network part of the model
def forward(self, img: Tensor,
        data_samples: Optional[List[TextDetDataSample]] = None
    ) -> Tuple[Tensor, Tensor, Tensor]:

# (Training) Step 2: DBModuleLoss, input the output of the DBNet network part and data_samples
def forward(self, preds: Tuple[Tensor, Tensor, Tensor],
                data_samples: Sequence[TextDetDataSample]) -> Dict:

# (Test) Step 2: DBPostProcessor
def get_text_instances(self, pred_results: Tuple[Tensor, Tensor, Tensor],
                           data_sample: TextDetDataSample
                           ) -> TextDetDataSample:
```

> As you can see from the above example, [`TextDetDataSample`](mmocr.structures.TextRecogDataSample) is used throughout the training and testing process of the detection model, and it encapsulates the data needed for the whole process of the text detection task.

Reviewer: Del, as it overlaps with the DataFlow part

> ### Text Recognition Task Data Abstraction TextRecogDataSample

Reviewer: ```suggestion
### Text Recognition - TextRecogDataSample
```

> [`TextRecogDataSample`](mmocr.structures.TextRecogDataSample) is used to encapsulate the data for the text recognition task. It has two properties, `gt_text` and `pred_text` , which are used to store prediction results and annotation information respectively.

Reviewer: fix the link

Reviewer: ```suggestion
[`TextRecogDataSample`](mmocr.structures.TextRecogDataSample) is used to encapsulate the data for the text recognition task. It has two fields, `gt_text` and `pred_text`, which are used to store annotation information and prediction results respectively.
```

Reviewer: fix CN version as well

|           |                           |                        |
| --------- | ------------------------- | ---------------------- |
| Field     | Type                      | Description            |
| gt_text   | [`LabelData`](#labeldata) | Label information.     |

> | pred_text | [`LabelData`](#labeldata) | The prediction result. |

Reviewer: fix links

Reviewer: ```suggestion
| gt_text   | [`LabelData`](#labeldata) | Label information.     |
| pred_text | [`LabelData`](#labeldata) | Prediction results. |
```

The following sample code demonstrates the use of [`TextRecogDataSample`](mmocr.structures.TextRecogDataSample).

```python
import torch
from mmengine.data import TextRecogDataSample

data_sample = TextRecogDataSample()
# Define the ground truth data
img_meta = dict(img_shape=(800, 1196, 3), pad_shape=(800, 1216, 3))
gt_text = LabelData(metainfo=img_meta)
gt_text.item = 'mmocr'
data_sample.gt_text = gt_text

# Define the prediction data
pred_text = LabelData(metainfo=img_meta)
pred_text.item = 'mmocr'
data_sample.pred_text = pred_text
```

> The fields of `Labelata` that will be used are:

Reviewer: ```suggestion
The fields of `Labeldata` that will be used are:
```

Reviewer: fix CN

|       |                     |                                                                                           |
| ----- | ------------------- | ----------------------------------------------------------------------------------------- |
| Field | Type                | Description                                                                               |
| item  | `list[str]`         | The text corresponding to the instance, of length (N, ), for end-to-end OCR tasks and KIE |

> | score | `torch.FloatTensor` | Confidence of the text prediction, of length (N, ), for the end-to-end OCR task           |

Reviewer: Missing `indexes` and `padded_indexes`
https://github.com/open-mmlab/mmocr/blob/8483b176b2d268303495666791a5a0e1535777b8/mmocr/models/textrecog/module_losses/base.py#L129-L130

Similarly, [`TextRecogDataSample`](mmocr.structures.TextRecogDataSample) runs through the entire training and testing process of the recognition model, as follows.

```python
# Encoder
def forward(self, feature: torch.Tensor,
                data_samples: List[TextRecogDataSample]) -> torch.Tensor:

# Decoder
def forward_train(self,
                      feat: Optional[torch.Tensor] = None,
                      out_enc: torch.Tensor = None,
                      data_samples: Sequence[TextRecogDataSample] = None
                      ) -> torch.Tensor:

# Module Loss
def forward(self, outputs: torch.Tensor,
                data_samples: Sequence[TextRecogDataSample]) -> Dict:

# Post Processor
def get_single_prediction(
        self,
        probs: torch.Tensor,
        data_sample: Optional[TextRecogDataSample] = None,
    ) -> Tuple[Sequence[int], Sequence[float]]:

> ```

Reviewer: del

> ### Key Information Extraction Task Data Abstraction KIEDataSample

Reviewer: ```suggestion
### Key Information Extraction - KIEDataSample
```

> [`KIEDataSample`](mmocr.structures.KIEDataSample) is used to encapsulate the data needed for the KIE task. It also agrees on two properties, `gt_instances` and `pred_instances`, which are used to store annotation information and prediction results respectively.

Reviewer: fix link

> [`KIEDataSample`](mmocr.structures.KIEDataSample) is used to encapsulate the data needed for the KIE task. It also agrees on two properties, `gt_instances` and `pred_instances`, which are used to store annotation information and prediction results respectively.

Reviewer: ```suggestion
[`KIEDataSample`](mmocr.structures.KIEDataSample) is used to encapsulate the data needed for the KIE task. It also contains two fields, `gt_instances` and `pred_instances`, which are used to store annotation information and prediction results respectively.
```

|                |                                 |             |
| -------------- | ------------------------------- | ----------- |
| Field          | Type                            | Description |

> | gt_instances   | [`InstanceData`](#instancedata) | Annotation  |
> | pred_instances | [`InstanceData`](#instancedata) | Prediction  |
> 

Reviewer: ```suggestion
| gt_instances   | [`InstanceData`](#instancedata) | Annotation information. |
| pred_instances | [`InstanceData`](#instancedata) | Prediction results.  |

```
To be consistent

> The [`InstanceData`](#instancedata) fields that will be used by this task are shown in the following table.

Reviewer: fix link

|             |                         |                                                                                                                                                                        |
| ----------- | ----------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Field       | Type                    | Description                                                                                                                                                            |

> | bboxes      | `torch.Tensor(float32)` | Bounding boxes `[x1, x2, y1, y2]` with the shape `(N, 4)`.                                                                                                             |

Reviewer: ```suggestion
| bboxes      | `torch.FloatTensor` | Bounding boxes `[x1, x2, y1, y2]` with the shape `(N, 4)`.                                                                                                             |
```

Reviewer: Same for CN

> | labels      | `torch.LongTensor`      | Instance label with the shape `(N, )`. By default, MMOCR uses `0` to represent the "text" class.                                                                       |

Reviewer: ```suggestion
| labels      | `torch.LongTensor`      | Instance label with the shape `(N, )`. |
```

| texts       | `list[str]`             | The textual content of each instance with the shape `(N, )`，used for e2e text spotting or KIE task.                                                                   |

> | edge_labels | `torch.IntTensor`       | The adjacency matrix between nodes with the shape `(N, N)`. In the KIE task, the optional values for the state between nodes are `-1` (ignored, not involved in loss calculation)，`0` (disconnected) and `1`(connected). |

Reviewer: ```suggestion
| edge_labels | `torch.IntTensor`       | The node adjacency matrix with the shape `(N, N)`. In the KIE task, the optional values for the state between nodes are `-1` (ignored, not involved in loss calculation)，`0` (disconnected) and `1`(connected). |
```

> | edge_scores | `torch.FloatTensor`     | The prediction confidence of each edge in the KIE task, with the shape `(N, N)`.                                                                                       |

Reviewer: Missing `scores`:
```
| scores | `torch.FloatTensor` | The confidence scores for node label predictions, with the shape `(N,)`. |
```

> Since there is no unified standard for model implementation of KIE tasks, the design currently considers only [SDMGR](../../../configs/kie/sdmgr/README.md) model usage scenarios. Therefore, the design is subject to change as we support more KIE models.

Reviewer: ````suggestion
```{warning}
Since there is no unified standard for KIE model implementation, the design currently considers only [SDMGR](../../../configs/kie/sdmgr/README.md) model usage scenarios. Therefore, the design is subject to change as we support more KIE models.
```
````

Reviewer: fix CN as well
